웹소켓 연동

apptie opened the pull request:
## 📎 관련 Issue 번호
<!-- (필수) 해당 PR과 관련 있는 issue 번호를 입력해주세요. -->
- closed #5 
## 📄 작업 내용 요약
<!-- (필수) 작업한 내용을 간단히 요약해주세요. -->
### Actor 설정 
- GuardianActor가 AbstractBehavior를 확장하지 않고 SpawnProtocol.create()를 반환하도록 변경 
- ClusterSharding에 ChatChannelEntity 초기화 로직 추가 

### 웹소켓 연동 
- 웹소켓 연동을 위한 컴포넌트 추가
  - WebSocketHandler 구현체 추가 
    - Actor 구조와의 연동 수행
  - ClientMessageSender 구현체 추가
    - 불필요한 close() 메서드 삭제
- 로컬 환경에서의 웹소켓 설정 추가
  - 추후 배포 시 운영 환경에서의 설정 추가 예정

apptie committed:
fix: 불필요한 로직 삭제

- Spring WebFlux에서 WebSocketSession 관리는 WebSocketHandler 구현체에서 수행

apptie committed:
feat: Cluster Entity Shard 초기화 로직 추가

- ChatChannelEntity 초기화 로직 추가

apptie committed:
feat: ClientMessageSender 웹소켓 구현체 추가

apptie committed:
feat: WebSocketHandler 구현체 추가

apptie committed:
fix: GuardianActor가 SpawnProtocol.Comman Behavior를 반환하도록 변경

- AbstractBehavior 상속 삭제

apptie committed:
chore: 웹소켓 설정 추가

- 로컬 실행을 위한 정적 리소스 client.html 추가

seonWKim commented on pekko/src/main/java/com/tok/pekko/global/config/websocket/WebSocketConfig.java:
```diff
@@ -0,0 +1,68 @@
+import org.springframework.web.reactive.socket.WebSocketHandler;
+import org.springframework.web.reactive.socket.server.support.WebSocketHandlerAdapter;
+
+@Profile("dev")
```
dev에만 동작하는 이유가 궁금해요? prod는 추가할 필요 없나요?

apptie replied:
당분간 배포를 할 계획이 없으며 프로젝트에 사용하는 기술 모두 처음 사용하는 기술이기 때문에 추가적인 고려사항이 있는 prod는 추후 배포 관련 이슈가 있을 때 진행하는 것이 생산성이 높다고 판단했습니다

seonWKim commented on pekko/src/main/java/com/tok/pekko/global/config/websocket/WebSocketConfig.java:
```diff
@@ -0,0 +1,68 @@
+        return new WebSocketHandlerAdapter();
+    }
+
+    private CorsConfiguration createCorsConfiguration() {
```
.yaml 파일 설정으로 뺄 수 있지 않을까요?

apptie replied:
profile로 환경별 분리까지 했는데 그 생각을 못했었네요
말씀해주신대로 간단한 설정이기도 하니 .yml로 분리하도록 하겠습니다

seonWKim commented on pekko/src/main/java/com/tok/pekko/adapter/in/websocket/DevChatWebSocketHandler.java:
```diff
@@ -0,0 +1,143 @@
+                                                                       .getQueryParams();
+
+            Long channelId = extractRequiredParam(params, "channelId");
+            Long userId = extractRequiredParam(params, "userId");
```
url으로부터 userId를 추출하면 보안에 취약하지는 않을까요? 

누구든 임의의 userId로 세션을 생성할 수 있는 구조 같아서요.

apptie replied:
그에 대해서는 추후 토큰 기반 인증으로 변경해 대응하고자 했습니다
특히 말씀해주신것처럼 서비스 자체 인증 기능을 사용하면 보안을 위해 너무 고려해야할게 많아진다고 생각해 oidc 기반의 소셜 로그인으로 진행할 생각을 가지고 있습니다

지금 코드는 dev기도 하고 해당 과정은 별도의 이슈로 다뤄야 할 만큼 큰 규모라고 생각해 지금처럼 보안에 취약하지만 간단한 방식을 선택했습니다

apptie replied:
이 문제는 제가 ChatWebSocketHandler의 사용 범위를 명확하게 표현하지 않아서 발생한 문제인 것 같습니다 
좀 더 명확하게 dev에서만 사용한다는 것을 표현하도록 하겠습니다

seonWKim commented on pekko/src/main/java/com/tok/pekko/adapter/in/websocket/ChatWebSocketHandler.java:
```diff
@@ -0,0 +1,143 @@
+        SendMessage command = new SendMessage(
+                context.userId(),
+                message.getPayloadAsText(),
+                LocalDateTime.now()
```
`now()` 대신 `now(ZoneId)`를 사용하는게 안전해요! 왜 그런지는 살펴보시면 좋을 것 같아요

apptie replied:
관련해서 찾아보도록 하겠습니다!
감사합니다

seonWKim commented on pekko/src/main/java/com/tok/pekko/adapter/out/websocket/ClientSessionActor.java:
```diff
@@ -45,8 +45,6 @@ private Behavior<ClientSessionCommand> onDeliverHistory(DeliverHistory command)
     }
 
     private Behavior<ClientSessionCommand> onShutdown(Shutdown command) {
```
`clientMessageSender.close()`는 왜 제거됐나요?

apptie replied:
이전에는 mvc 방식처럼 생각해 close를 외부에서 명시적으로 호출하기 위해 정의했었습니다

Webflux에서는 inbound, outbound 메시지를 WwbsocketHandler.handle()에서 설정하고 Mono<Void>를 반환해 웹소켓 세션의 생명주기를 관리하므로 명시적으로 외부에서 close를 하지 않아도 된다고 판단했습니다

apptie committed:
refactor: Clock 의존성 주입으로 타임존 설정 지원

- @ConfigurationProperties를 활용한 Timezone 설정

apptie committed:
refactor: WebSocket 설정을 @ConfigurationProperties로 외부화

- @Profile("dev")를 제거해 모든 환경에서 사용 가능하도록 변경
- WebSocketProperties 추가로 profile별 설정 관리

apptie committed:
refactor: WebSocketHandler를 dev 환경 전용으로 분리

- ChatWebSocketHandler를 DevChatWebSocketHandler로 이름 변경
- @Profile("dev") 추가하여 dev 환경에서만 활성화
